src/magicbox/QueriesProcessor.ts

```diff
@@ -0,0 +1,91 @@
+import { defaults } from 'underscore';
+
+export interface IQueriesProcessorOptions {
+  timeout: number;
+}
+
+export enum QueryProcessResultStatus {
+  Finished = 'finished',
+  TimedOut = 'timedout',
+  Overriden = 'overriden'
+}
+
+export type QueryProcessResult<ItemType> =
+  | {
+      status: QueryProcessResultStatus.Finished | QueryProcessResultStatus.TimedOut;
+      items: ItemType[];
+    }
+  | {
+      status: QueryProcessResultStatus.Overriden;
+    };
+
+export class QueriesProcessor<ItemType> {
+  private static async streamQueries<ItemType>(queries: Promise<ItemType[]>[], output: ItemType[]) {
+    await Promise.all(queries.map(query => query.then(items => output.push(...items))));
+  }
+
+  private override: () => any;
+  private options: IQueriesProcessorOptions;
+
+  constructor(options: Partial<IQueriesProcessorOptions> = {}) {
```

Comment:
Here we have a relatively complex constructor, using complex type system, with partials, with default parameter value, which then is merged with a default object.

But, at the end of the day, there's a single option value, with a single parameter, which is the timeout.

Why not : 

```
constructor(private timeout = 500)
```

?

Reply:
True it is only one option now, but I prefer an object since it is easier to add more constructor args in the future.

Reply:
I'll leave it like that until there's an agreement on what to do with it.